### PR TITLE
Feature/new track properties

### PR DESCRIPTION
**Pull request by cclaessens**
This will solve the issues #134 #130 #138

Regarding issue #135:
I did not remove the slot that takes discriminated points and the power spectrum slice to sum over adjacent bins. Instead I added a second slot that only takes discriminated points.

Tested with TestSequentialTrackFinder.cc

### CODE REVIEW

File: Source/Data/SpectrumAnalysis/KTSequentialLineData.cc [hunk 1]
```diff
@@ -0,0 +1,126 @@
+/*
```
Comment (guiguem) [inline Source/Data/SpectrumAnalysis/KTSequentialLineData.cc]: This should look like 
```
/**
 @file KTTrackProcessing.hh
 @brief Contains KTTrackProcessing
 @details Extracts physics-relevant information about tracks
 @author: N.S. Oblath & B. LaRoque
 @date: July 22, 2013
 */
```

File: Source/Data/SpectrumAnalysis/KTSequentialLineData.hh [hunk 2]
```diff
@@ -0,0 +1,93 @@
+/*
```
Comment (guiguem) [inline Source/Data/SpectrumAnalysis/KTSequentialLineData.hh]: Should look like: 
```
/**
 @file KTTrackProcessing.hh
 @brief Contains KTTrackProcessing
 @details Extracts physics-relevant information about tracks
 @author: N.S. Oblath & B. LaRoque
 @date: July 22, 2013
 */
```

File: Source/Data/SpectrumAnalysis/KTSequentialLineData.hh [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+            MEMBERVARIABLE(double, AmplitudeSum);
+            MEMBERVARIABLE(double, SNRSum);
+            MEMBERVARIABLE(double, NUPSum);
+            MEMBERVARIABLE(unsigned, NPoints);
+            MEMBERVARIABLE(uint64_t, AcquisitionID);
+            MEMBERVARIABLE(unsigned, MinPoints);
+            MEMBERVARIABLE(double, StartFrequencySigma);
+            MEMBERVARIABLE(double, EndFrequencySigma);
+
+            double fSumX;
```
Comment (guiguem) [inline Source/Data/SpectrumAnalysis/KTSequentialLineData.hh]: private variables?

File: Source/Data/SpectrumAnalysis/KTSequentialLineData.hh [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+            double fSumY;
+            double fSumXY;
+            double fSumXX;
+
+            KTSequentialLineData();
```
Comment (guiguem) [inline Source/Data/SpectrumAnalysis/KTSequentialLineData.hh]: Public variables again...

File: Source/Data/SpectrumAnalysis/KTSequentialLineData.cc [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+        SetEndTimeInAcq( fLinePoints.rbegin()->fTimeInAcq);
+        SetEndFrequency( fLinePoints.rbegin()->fFrequency);
+    }
+
+
+    inline void KTSequentialLineData::UpdateLineProperties()
+    {
+        //KTDEBUG(seqlog, "Updating line parameters");
+        SetNPoints( fLinePoints.size() );
+        if ( fNPoints == 1 )
```
Comment (guiguem) [inline Source/Data/SpectrumAnalysis/KTSequentialLineData.cc]: Why two cases here for the same action?

File: Source/EventAnalysis/KTOverlappingTrackClustering.cc [hunk 6]
```diff
@@ -92,22 +134,44 @@ namespace Katydid
 
     bool KTOverlappingTrackClustering::Run()
     {
-        return DoClustering();
+        if (fCompTracks.size() != 0 )
```
Comment (guiguem) [inline Source/EventAnalysis/KTOverlappingTrackClustering.cc]: Shouldn't you decide on using one or the other depending on the type of slot you got and if the number of CompTrack is indeed non zero?

File: Source/EventAnalysis/KTOverlappingTrackClustering.cc [hunk 7]
```diff
@@ -92,22 +134,44 @@ namespace Katydid
...
+        {
+            KTINFO( otclog, "Clustering procTracks");
```
Comment (guiguem) [inline Source/EventAnalysis/KTOverlappingTrackClustering.cc]: `Clustering ProcessedTracks`

File: Source/EventAnalysis/KTOverlappingTrackClustering.cc [hunk 8]
```diff
@@ -92,22 +134,44 @@ namespace Katydid
...
+            return DoTrackClustering();
+        }
+        else
+        {
+            KTINFO( otclog, "Clustering SeqLine Candidates");
```
Comment (guiguem) [inline Source/EventAnalysis/KTOverlappingTrackClustering.cc]: `Clustering Sequential Line`

File: Source/EventAnalysis/KTOverlappingTrackClustering.cc [hunk 9]
```diff
@@ -143,6 +207,40 @@ namespace Katydid
...
+            while (numberOfCands!=numberOfNewCands)
+            {
+                numberOfCands = fCompSeqLineCands.size();
+                KTDEBUG(otclog, "Number of SeqLine candidates to cluster: "<< numberOfCands);
+                this->OverlapSeqLineClustering();
+
+                // Update number of tracks
+                numberOfNewCands = fNewSeqLineCands.size();
+
+                KTDEBUG(otclog, "Number of new SeqLine candidates: "<< numberOfNewCands);
```
Comment (guiguem) [inline Source/EventAnalysis/KTOverlappingTrackClustering.cc]: `New number of SequentialLine Candidates`

File: Source/EventAnalysis/KTOverlappingTrackClustering.cc [hunk 10]
```diff
@@ -143,6 +207,40 @@ namespace Katydid
 
         return true;
     }
+    bool KTOverlappingTrackClustering::FindMatchingSeqLineCands()
```
Comment (guiguem) [inline Source/EventAnalysis/KTOverlappingTrackClustering.cc]: Template methods are a thing!
Maybe you can use that to simplify this file a bit.

File: Source/Data/SpectrumAnalysis/KTSequentialLineData.cc [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+#include <cmath>
+#include <algorithm>
+
+namespace Katydid
+{
+    const std::string KTSequentialLineData::sName("sequential-line");
+	KTLOGGER(seqlog, "KTSeqLine");
+
+    KTSequentialLineData::KTSequentialLineData():
+        fStartTimeInRunC(0.0),
```
Comment (nsoblath) [inline Source/Data/SpectrumAnalysis/KTSequentialLineData.cc]: Some variables are not initialized.

File: Source/Data/SpectrumAnalysis/KTSequentialLineData.hh [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+#include <set>
+
+namespace Katydid
+{
+    class KTSequentialLineData : public Nymph::KTExtensibleData< KTSequentialLineData >
+    {
+
+        private:
+
+            KTDiscriminatedPoints fLinePoints;
```
Comment (nsoblath) [inline Source/Data/SpectrumAnalysis/KTSequentialLineData.hh]: MEMBERVARIABLE macros should be used here too

File: Source/EventAnalysis/KTOverlappingTrackClustering.hh [hunk 13]
```diff
@@ -56,46 +70,191 @@ namespace Katydid
...
+            }
 
-        private:
-            bool OverlapClustering();
-            bool DoTheyOverlap(KTProcessedTrackData& track1, KTProcessedTrackData& track2);
-            bool DoTheyCross(KTProcessedTrackData& track1, KTProcessedTrackData& track2);
-            const void CombineTracks(const KTProcessedTrackData& track1, KTProcessedTrackData& track2);
-            bool FindMatchingTracks();
-            void EmitTrackCandidates();
+            template<typename T>
```
Comment (nsoblath) [inline Source/EventAnalysis/KTOverlappingTrackClustering.hh]: For all of these templated functions:
* Please move the function implementations outside of the class definition.  You can put it in this file, below the class definition.
* Please use a more descriptive type name than `T`.

File: Source/Data/SpectrumAnalysis/KTSequentialLineData.hh (hunk 4 above)
Comment (nsoblath) [reply Source/Data/SpectrumAnalysis/KTSequentialLineData.hh]: I second Mathieu's comment.  The variables should be private with any necessary access functions provided.  Feel free to use a MEMBERVARIABLE macro if appropriate.

File: Source/EventAnalysis/KTOverlappingTrackClustering.hh [hunk 14]
```diff
@@ -15,8 +15,11 @@
 #include "KTData.hh"
 #include "KTMemberVariable.hh"
 #include "KTProcessedTrackData.hh"
+#include "KTSequentialLineData.hh"
+#include "KTDiscriminatedPoint.hh"
 
 #include <vector>
+#include <iostream>
```
Comment (nsoblath) [inline Source/EventAnalysis/KTOverlappingTrackClustering.hh]: You shouldn't need to include `iostream`

File: Source/Executables/Validation/CMakeLists.txt [hunk 15]
```diff
@@ -82,6 +82,7 @@ if (Katydid_ENABLE_TESTING)
         TestSpectrogramStriperSwaps
         TestSpectrumDiscriminator
         TestWindowFunction
+        TestSequentialTrackFinder
```
Comment (nsoblath) [inline Source/Executables/Validation/CMakeLists.txt]: These are listed alphabetically.  Please move it up a few lines to the appropriate place.

File: Source/Executables/Validation/TestSequentialTrackFinder.cc [hunk 16]
```diff
@@ -0,0 +1,240 @@
...
+#include "KTSequentialTrackFinder.hh"
+#include "KTOverlappingTrackClustering.hh"
+#include "KTIterativeTrackClustering.hh"
+#include "KTDiscriminatedPoints1DData.hh"
+#include "KTSequentialLineData.hh"
+#include "KTDiscriminatedPoint.hh"
+#include "KTRandom.hh"
+
+
+
```
Comment (nsoblath) [inline Source/Executables/Validation/TestSequentialTrackFinder.cc]: Global variables are bad even in a test program.  Please figure out how to make this work with the variables defined inside the `main()` function or in some other way.

File: Source/SpectrumAnalysis/KTSequentialTrackFinder.hh [hunk 17]
```diff
@@ -18,9 +18,11 @@
...
 #include "KTGainVariationData.hh"
-#include "KTSeqLine.hh"
-#include "KTProcessedTrackData.hh"
+#include "KTSequentialLineData.hh"
+//#include "KTProcessedTrackData.hh"
+#include "KTSparseWaterfallCandidateData.hh"
 #include "KTDiscriminatedPoints1DData.hh"
+#include "KTDiscriminatedPoint.hh"
 
 #include <iostream>
```
Comment (nsoblath) [inline Source/SpectrumAnalysis/KTSequentialTrackFinder.hh]: Shouldn't need to include `iostream`

File: Source/EventAnalysis/KTIterativeTrackClustering.hh [hunk 18]
```diff
@@ -88,33 +97,174 @@ namespace Katydid
             MEMBERVARIABLE(double, AverageUnitlessResidualThreshold);
             MEMBERVARIABLE(unsigned, NTracks);
 
+        private:
```
Comment (guiguem) [inline Source/EventAnalysis/KTIterativeTrackClustering.hh]: You should leave the methods declarations in the class, and move their definition outside at the bottom of the header file: this makes the class cleaner.

File: Source/Executables/Validation/TestSequentialTrackFinder.cc (hunk 16 above)
Comment (guiguem) [reply Source/Executables/Validation/TestSequentialTrackFinder.cc]: @cclaessens Have you solved this comment?

**Comment by guiguem** [comment]
I approved these changes but the question of using KTApplyCut instead of custom cuts remains open.

File: Source/Executables/Validation/TestSequentialTrackFinder.cc (hunk 16 above)
Comment (cclaessens) [reply Source/Executables/Validation/TestSequentialTrackFinder.cc]: yes I moved them inside the functions or passed them as function parameters

**Pull request merged**